### PR TITLE
fix(cc): Continue CCP calculation if cpu_cores < cu_number [fixes NET-808]

### PR DESCRIPTION
**Pull request by gurinderu**
## Description
Add a warning message if commitment > available host CPU count. + minor fixes

## Related Issue(s)
[NET-808](https://linear.app/fluence/issue/NET-808/continue-ccp-calculation-if-cpu-cores-cu-number)

## Checklist
- [ ] The code follows the project's coding conventions and style guidelines.
- [ ] All tests related to the changes have passed successfully.
- [ ] Documentation has been updated to reflect the changes (if applicable).
- [ ] All new and existing unit tests have passed.
- [ ] I have self-reviewed my code and ensured its quality.
- [ ] I have added/updated necessary comments to aid understanding.

## Reviewer Checklist
- [ ] Code has been reviewed for quality and adherence to [guidelines](https://doc.rust-lang.org/1.0.0/style/README.html).
- [ ] Tests have been reviewed and are sufficient to validate the changes.
- [ ] Documentation has been reviewed and is up to date.
- [ ] Any questions or concerns have been addressed.

### CODE REVIEW

File: crates/chain-listener/src/listener.rs [hunk 1]
```diff
@@ -794,46 +795,31 @@ impl ChainListener {
...
-    }
-
-    /// Return units in CC that is not active yet and can't produce proofs in the current epoch
-    fn get_pending_units(&self) -> Vec<CUID> {
-        self.cc_compute_units
-            .values()
-            .filter(|cu| cu.startEpoch > self.current_epoch) // CU hasn't yet started
-            .map(|cu| CUID::new(cu.id.0))
-            .collect()
+    fn group_cc_units(&self) -> CUGroupResult {
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:798]:
```suggestion
    fn get_cu_groups(&self) -> CUGroupResult {
```

or `get_cu_layout`

`group_cc_units` has the following problems: verb is `group`, implying function is `&mut`; `cc_units` term doesn't exist in our lingo.

File: crates/chain-listener/src/listener.rs [hunk 2]
```diff
@@ -794,46 +795,31 @@ impl ChainListener {
...
+                } else if *count >= self.max_proofs_per_epoch {
+                    finished_units.push(*cuid)
+                } else {
+                    non_priority_units.push(*cuid)
+                }
+            } else {
+                pending_units.push(*cuid);
+            }
+        }
+        CUGroupResult {
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:817]: I'd call it `CULayout` or `CUGroups`. `Result` is close to Result from std, which confuses a bit.

File: crates/chain-listener/src/listener.rs [hunk 3]
```diff
@@ -1225,13 +1273,27 @@ impl ChainListener {
...
+}
+
+struct PhysicalCoreGroupResult {
+    pub priority_cores: Vec<PhysicalCoreId>,
+    pub non_priority_cores: Vec<PhysicalCoreId>,
+    pub pending_cores: Vec<PhysicalCoreId>,
+    pub finished_cores: Vec<PhysicalCoreId>,
+}
+
 fn all_min_proofs_found(priority_units: &[CUID]) -> bool {
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:1297]: Maybe make it a method on `CUGroups`/`CULayout`?

Comment (folex) [reply crates/chain-listener/src/listener.rs:1297]: Then it would be `if cu_groups.all_min_proofs_found()` instead of `all_min_proofs_found(cu_groups)`

File: crates/chain-listener/src/listener.rs (hunk 1 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs]: Fixed

File: crates/chain-listener/src/listener.rs (hunk 2 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs]: Renamed

File: crates/chain-listener/src/listener.rs (hunk 3 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs]: done

File: crates/chain-listener/src/listener.rs [hunk 4]
```diff
@@ -878,44 +864,54 @@ impl ChainListener {
             None => return Ok(()),
         };
 
-        let priority_units = self.get_priority_units();
-        let non_priority_units = self.get_non_priority_units();
-        let pending_units = self.get_pending_units();
+        let group_result = self.get_cu_groups();
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:867]: why `_result`? Sounds like `groups`

File: crates/chain-listener/src/listener.rs [hunk 5]
```diff
@@ -1274,12 +1322,27 @@ impl ChainListener {
...
+impl CUGroups {
+    fn all_min_proofs_found(&self) -> bool {
+        self.priority_units.is_empty()
+    }
+
+    fn all_max_proofs_found(&self) -> bool {
+        self.non_priority_units.is_empty()
+    }
+}
+struct PhysicalCoreGroupResult {
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:1341]: Why `Result` suffix?

File: crates/chain-listener/src/listener.rs [hunk 6]
```diff
@@ -949,19 +945,71 @@ impl ChainListener {
...
-                units.to_vec(),
-                WorkType::CapacityCommitment,
-            ))
-            .map_err(|err| {
-                tracing::error!(target: "chain-listener", "Failed to acquire cores for active units: {err}");
-                eyre::eyre!("Failed to acquire cores for active units: {err}")
-            })?;
+    fn acquire_cores_for_cc(
+        &self,
+        cu_group_result: &CUGroups,
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:950]: `_result` let's remove

File: crates/chain-listener/src/listener.rs [hunk 7]
```diff
@@ -878,44 +864,54 @@ impl ChainListener {
...
+                cu_allocation.extend(
+                    cc_cores
+                        .non_priority_cores
+                        .iter()
+                        .cloned()
+                        .zip(group_result.non_priority_units.iter().cloned()),
+                );
+
+                let mut units = group_result.non_priority_units.iter().cycle();
                 // Assign "pending cores" to help generate proofs for "non priority units"
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:890]: It took me 10-20 minutes to understand what's happening here :( Names are super hard to understand.

```suggestion
                // Assign "pending cores" to help generate proofs for "non priority units"
                let mut units = group_result.non_priority_units.iter().cycle();
```

File: crates/chain-listener/src/listener.rs [hunk 8]
```diff
@@ -878,44 +864,54 @@ impl ChainListener {
...
                 });
             }
         } else {
             // Use assigned cores to calculate proofs for CUs who haven't reached MIN_PROOF_COUNT yet
-            cu_allocation.extend(priority_cores.iter().zip(priority_units.iter()));
+            cu_allocation.extend(
+                cc_cores
+                    .priority_cores
+                    .iter()
+                    .zip(group_result.priority_units.iter()),
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:903]: move to a separate variable please

File: crates/chain-listener/src/listener.rs [hunk 9]
```diff
@@ -878,44 +864,54 @@ impl ChainListener {
...
                 // All CUs were proven, now let's work on submitting proofs for every CU until MAX_PROOF_COUNT is reached
-                cu_allocation.extend(non_priority_cores.iter().zip(non_priority_units.iter()));
-
-                let mut units = non_priority_units.iter().cycle();
+                cu_allocation.extend(
+                    cc_cores
+                        .non_priority_cores
+                        .iter()
+                        .cloned()
+                        .zip(group_result.non_priority_units.iter().cloned()),
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:886]: let's move to a separate variable with a good name, please

File: crates/chain-listener/src/listener.rs [hunk 10]
```diff
@@ -949,19 +945,71 @@ impl ChainListener {
...
-                eyre::eyre!("Failed to acquire cores for active units: {err}")
-            })?;
+    fn acquire_cores_for_cc(
+        &self,
+        cu_group_result: &CUGroups,
+    ) -> eyre::Result<PhysicalCoreGroupResult> {
+        let mut units = cu_group_result.priority_units.clone();
+        units.extend(cu_group_result.non_priority_units.clone());
+        units.extend(cu_group_result.pending_units.clone());
+        units.extend(cu_group_result.finished_units.clone());
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:955]: clone clone clone clone

File: crates/chain-listener/src/listener.rs [hunk 11]
```diff
@@ -949,19 +945,71 @@ impl ChainListener {
...
+        ));
+
+        fn filter(units: &[CUID], assignment: &Assignment) -> Vec<PhysicalCoreId> {
+            units
+                .iter()
+                .filter_map(|cuid| {
+                    assignment
+                        .cuid_core_data
+                        .get(cuid)
+                        .map(|data| data.physical_core_id)
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:969]: separate variable please

File: crates/chain-listener/src/listener.rs [hunk 12]
```diff
@@ -949,19 +945,71 @@ impl ChainListener {
...
+                })
+            }
+            Err(AcquireError::NotFoundAvailableCores {
+                required,
+                available,
+                ..
+            }) => {
+                tracing::warn!("Found {required} CUs in the Capacity Commitment, but Nox has only {available} Cores available for CC");
+                let assign_units = units.iter().take(available).cloned().collect();
+                self.core_manager.release(units);
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:995]: This is a super important action, and it looks just a regular line in the code.

Add a comment, but better move all the effectful lines to the end of the function/block

File: crates/chain-listener/src/listener.rs [hunk 13]
```diff
@@ -949,19 +945,71 @@ impl ChainListener {
...
+                    assign_units,
+                    WorkType::CapacityCommitment,
+                ))?;
+                let priority_units = filter(&cu_group_result.priority_units, &assignment);
+                let non_priority_units = filter(&cu_group_result.non_priority_units, &assignment);
+                let pending_units = filter(&cu_group_result.pending_units, &assignment);
+                let finished_units = filter(&cu_group_result.finished_units, &assignment);
+
+                Ok(PhysicalCoreGroupResult {
+                    priority_cores: priority_units,
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:1006]: Why do you assign something called `*_units` to something called `*_cores`? `cores: units` – this blows my mind

File: crates/core-manager/src/dev.rs [hunk 14]
```diff
@@ -240,21 +236,19 @@ impl CoreManagerFunctions for DevCoreManager {
         let mut lock = self.state.write();
         let mut result_physical_core_ids = BTreeSet::new();
         let mut result_logical_core_ids = BTreeSet::new();
+        let mut cuid_core_data: Map<CUID, CoreData> = HashMap::with_capacity_and_hasher(
+            assign_request.unit_ids.len(),
+            FxBuildHasher::default(),
```
Comment (folex) [inline crates/core-manager/src/dev.rs:241]: Why do you need a custom hasher? Please leave a comment in code

File: crates/core-manager/src/dev.rs [hunk 15]
```diff
@@ -240,21 +236,19 @@ impl CoreManagerFunctions for DevCoreManager {
...
-                            .map(|(k, v)| (*k, *v))
-                            .collect();
-                        AcquireError::NotFoundAvailableCores {
-                            current_assignment: CurrentAssignment::new(current_assignment),
-                        }
-                    })?;
+                    let core_id = lock
+                        .available_cores
+                        .pop_front()
+                        .expect("Unexpected state. Should not be empty never");
```
Comment (folex) [inline crates/core-manager/src/dev.rs:251]: Please add `// SAFETY` explanation on why this would never happen.

I would prefer to remove this `expect`, and instead do matching on `Option`, and report to logs if that happens.

File: crates/core-manager/src/dev.rs [hunk 16]
```diff
@@ -240,21 +236,19 @@ impl CoreManagerFunctions for DevCoreManager {
         let mut lock = self.state.write();
         let mut result_physical_core_ids = BTreeSet::new();
         let mut result_logical_core_ids = BTreeSet::new();
+        let mut cuid_core_data: Map<CUID, CoreData> = HashMap::with_capacity_and_hasher(
```
Comment (folex) [inline crates/core-manager/src/dev.rs:239]: what does `_data` mean? `core_data` – what is that?

File: crates/core-manager/src/types.rs [hunk 17]
```diff
@@ -24,10 +25,17 @@ impl AcquireRequest {
     }
 }
 
-#[derive(Debug, Eq, PartialEq)]
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct CoreData {
```
Comment (folex) [inline crates/core-manager/src/types.rs:29]: What is `Data`? It's super generic, could mean literally anything. It's like `CoreThings`. Could you come up with something more specific, please?

**Review by folex — COMMENTED** [review]
There are so many clones in `refresh_commitment`, it makes it super hard to follow the logic.

File: crates/chain-listener/src/listener.rs [hunk 18]
```diff
@@ -813,46 +814,31 @@ impl ChainListener {
         Ok(())
     }
 
-    /// Return already started units involved in CC and not having less than MIN_PROOFS_PER_EPOCH proofs in the current epoch
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:816]: you have removed documentation from code, but didn't add any.

File: crates/chain-listener/src/listener.rs [hunk 19]
```diff
@@ -813,46 +814,31 @@ impl ChainListener {
...
-                    .get(cuid)
-                    .map(|count| *count < self.min_proofs_per_epoch)
-                    .unwrap_or(true)
-            })
-            .map(|(cuid, _)| *cuid)
-            .collect()
-    }
-
-    /// Return already started units involved in CC and found at least MIN_PROOFS_PER_EPOCH proofs,
-    /// but less that MAX_PROOFS_PER_EPOCH proofs in the current epoch
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:832]: same

File: crates/chain-listener/src/listener.rs (hunk 4 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs]: done

File: crates/chain-listener/src/listener.rs (hunk 5 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs]: done

File: crates/chain-listener/src/listener.rs (hunk 6 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs]: Done

File: crates/chain-listener/src/listener.rs (hunk 7 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs]: done

File: crates/chain-listener/src/listener.rs (hunk 13 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs]: done

File: crates/core-manager/src/dev.rs (hunk 15 above)
Comment (gurinderu) [reply crates/core-manager/src/dev.rs:253]: done

File: crates/core-manager/src/dev.rs (hunk 16 above)
Comment (gurinderu) [reply crates/core-manager/src/dev.rs]: renamed to cores

File: crates/core-manager/src/types.rs (hunk 17 above)
Comment (gurinderu) [reply crates/core-manager/src/types.rs]: renamed to Cores

File: crates/core-manager/src/dev.rs (hunk 14 above)
Comment (gurinderu) [reply crates/core-manager/src/dev.rs:242]: added comment in the Assignment struct

File: crates/chain-listener/src/listener.rs (hunk 18 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs:816]: Added into CUGroups

File: crates/chain-listener/src/listener.rs (hunk 19 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs:832]: Added into CUGroups

File: crates/chain-listener/src/listener.rs [hunk 20]
```diff
@@ -813,46 +814,31 @@ impl ChainListener {
...
-                    .map(|count| {
-                        *count >= self.min_proofs_per_epoch && *count < self.max_proofs_per_epoch
-                    })
-                    .unwrap_or(true)
-            })
-            .map(|(cuid, _)| *cuid)
-            .collect()
-    }
-
-    /// Return units in CC that is not active yet and can't produce proofs in the current epoch
```
[the remark this answers is not in the document]
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs:849]: Added into CUGroups

File: crates/chain-listener/src/listener.rs (hunk 11 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs]: done

File: crates/chain-listener/src/listener.rs (hunk 9 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs]: done

File: crates/chain-listener/src/listener.rs (hunk 8 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs]: done

File: crates/chain-listener/src/listener.rs (hunk 12 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs:991]: Done

File: crates/chain-listener/src/listener.rs (hunk 10 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs]: fixed

File: crates/core-manager/src/dev.rs [hunk 21]
```diff
@@ -127,6 +127,7 @@ impl DevCoreManager {
 
         let mut system_cores: BTreeSet<PhysicalCoreId> = BTreeSet::new();
         for _ in 0..system_cpu_count {
+            // SAFETY: this should never happen because we already checked the availability of cores
```
Comment (folex) [inline crates/core-manager/src/dev.rs:130]: Could you please show me, where do you check this?

File: crates/core-manager/src/strict.rs [hunk 22]
```diff
@@ -223,7 +224,7 @@ impl CoreManagerFunctions for StrictCoreManager {
         assign_request: AcquireRequest,
     ) -> Result<Assignment, AcquireError> {
         let mut lock = self.state.write();
-        let mut cuid_core_data: Map<CUID, CoreData> = HashMap::with_capacity_and_hasher(
+        let mut cuid_core_data: Map<CUID, Cores> = HashMap::with_capacity_and_hasher(
```
Comment (folex) [inline crates/core-manager/src/strict.rs:227]: _data

File: crates/chain-listener/src/listener.rs [hunk 23]
```diff
@@ -864,30 +864,29 @@ impl ChainListener {
...
+                    .iter()
+                    .cloned()
+                    .zip(cu_groups.non_priority_units.iter().cloned());
+                cu_allocation.extend(non_priority_cores_mapping);
+
                 // Assign "pending cores" to help generate proofs for "non priority units"
+                let mut units = cu_groups.non_priority_units.iter().cycle();
                 cc_cores.pending_cores.iter().for_each(|core| {
                     if let Some(unit) = units.next() {
                         cu_allocation.insert(*core, *unit);
```
Comment (folex) [inline crates/chain-listener/src/listener.rs:892]:
```suggestion
                let mut units = cu_groups.non_priority_units.iter().cycle();
                cc_cores.pending_cores.iter().for_each(|pending_core| {
                    if let Some(non_priority_unit) = units.next() {
                        cu_allocation.insert(*pending_core, *non_priority_unit);
```

**Review by folex — APPROVED** [review]
Thank you!

File: crates/core-manager/src/dev.rs (hunk 21 above)
Comment (gurinderu) [reply crates/core-manager/src/dev.rs:130]: line 93

File: crates/core-manager/src/strict.rs (hunk 22 above)
Comment (gurinderu) [reply crates/core-manager/src/strict.rs]: Fixed

File: crates/chain-listener/src/listener.rs (hunk 23 above)
Comment (gurinderu) [reply crates/chain-listener/src/listener.rs]: Fixed
